File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/medium/MaximumStarSumOfAGraph/maximum-star-sum-of-a-graph-ye15.py

from typing import List
from math import inf
class S:
    def maxStarSum(self, vals: List[int], edges: List[List[int]], k: int) -> int:
        
        n = len(vals)
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        ans = -inf
        for i, u in enumerate(graph):
            u.sort(key=vals.__getitem__, reverse=True)
            cand = vals[i] + sum(max(0, vals[x]) for x in u[:k])
            ans = max(ans, cand)
        
        return ans 
    
vals = [1,2,3,4,10,-10,-20]
edges = [[0,1],[1,2],[1,3],[3,4],[3,5],[3,6]]
k = 2
    
print(S().maxStarSum(vals, edges, k))

File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/leetcode/medium/MaximumStarSumOfAGraph/MaximumStarSumOfAGraphYe15.java
package com.mycompany.leetcode.medium.MaximumStarSumOfAGraph;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class MaximumStarSumOfAGraphYe15 {
    static int maxStarSum(int[] vals, int[][] edges, int k) {
        int n = vals.length;
        List<Integer>[] graph = new ArrayList[n];
        for(int u = 0; u < n; ++u) graph[u] = new ArrayList<>();
        for(int[] e: edges) {
            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }

        int ans = Integer.MIN_VALUE;
        for(int u = 0; u < n; ++u) {
            int cand = vals[u];
            if(graph[u].size() > k) Collections.sort(graph[u], (a, b) -> vals[b] - vals[a]);
            for(int v = 0; v < k && v < graph[u].size(); ++v)
                cand += Math.max(0, vals[graph[u].get(v)]);

            ans = Math.max(ans, cand);
        }

        return ans;
    }

    public static void main(String[] args) {
        int[] vals = {1,2,3,4,10,-10,-20};
        int[][] edges = {{0,1},{1,2},{1,3},{3,4},{3,5},{3,6}};
        int k = 2;

        System.out.println(maxStarSum(vals, edges, k));
    }
}
